### PR TITLE
feat: add gpu support to executors

### PR DESCRIPTION
**Pull request by jakob1111996**
This PR adds GPU support and tests to compatible executors.

### CODE REVIEW

File: jinahub/encoders/image/ImageTFEncoder/Dockerfile.cuda113 [hunk 1]
```diff
@@ -0,0 +1,10 @@
+FROM nvidia/cuda:11.3.1-cudnn8-runtime
+
+RUN apt update && apt install -y python3.8 python3-pip git
```
Comment (numb3r3) [inline jinahub/encoders/image/ImageTFEncoder/Dockerfile.cuda113:3]:
```suggestion
RUN apt update && apt install -y git python3.8 python3-pip \
    && cd /usr/local/bin \
    && ln -nsf /usr/bin/python3.8 python \
    && python -m pip install --no-cache-dir --upgrade pip
    
RUN JINA_PIP_INSTALL_PERF=1 pip install --no-cache-dir jina==2.0.20
```

File: jinahub/encoders/text/TransformerTorchEncoder/Dockerfile.cuda113 [hunk 2]
```diff
@@ -1,11 +1,11 @@
 FROM nvidia/cuda:11.3.1-cudnn8-runtime
 
-RUN apt update && apt install -y python3.8 python3-pip
+RUN apt update && apt install -y python3.8 python3-pip git
```
Comment (numb3r3) [inline jinahub/encoders/text/TransformerTorchEncoder/Dockerfile.cuda113:3]:
```suggestion
RUN apt update && apt install -y git python3.8 python3-pip \
    && cd /usr/local/bin \
    && ln -nsf /usr/bin/python3.8 python \
    && python -m pip install --no-cache-dir --upgrade pip
    
RUN JINA_PIP_INSTALL_PERF=1 pip install --no-cache-dir jina==2.0.20
```

File: jinahub/encoders/text/TransformerTorchEncoder/Dockerfile.cuda113 [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
 
 COPY . ./transformer-text-encoder/
 WORKDIR ./transformer-text-encoder
 
 RUN pip install jina[perf]>=2.0
```
Comment (numb3r3) [inline jinahub/encoders/text/TransformerTorchEncoder/Dockerfile.cuda113:8]:
```suggestion
```

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -15,7 +15,7 @@ jobs:
           sudo apt-get install jq
           echo "::set-output name=matrix::$(bash scripts/get-all-test-paths.sh)"
         env:
-          CHANGED_FILES: ${{ steps.changed-files.outputs.added_modified }}
+          CHANGED_FILES: ${{ steps.changed-files.outputs.all }}
```
Comment (cristianmtr) [inline .github/workflows/ci.yml:18]: Why did you change this?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -63,8 +63,45 @@ jobs:
       - name: run integration tests
         run: ./scripts/integration.sh
 
+  gpu-containers:
+    runs-on: [ self-hosted, gpu ]
```
Comment (cristianmtr) [inline .github/workflows/ci.yml:67]: self-hosted? Maybe i'm not up to date with this task. Do we have self-hosted runners?

Comment (tadejsv) [reply .github/workflows/ci.yml:67]: Yes. Right now a test one (although it is full featured), after this PR I will create a scalable deployment. Go to `Settings > Actions > Runners` in this repo to see it :)

**Comment by cristianmtr** [comment]
Make sure to also test the docker image as I am doing in `docker-tests`

File: .github/workflows/ci.yml (hunk 4 above)
Comment (jakob1111996) [reply .github/workflows/ci.yml:18]: Because I noticed that half of the tests did not run (for example if I delete/rename stuff that is breaking something nobody would notice.)

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -63,8 +63,58 @@ jobs:
...
+      fail-fast: false
+      matrix:
+        folder: ${{fromJson(needs.gpu-folders.outputs.matrix)}}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: debug
+        run: echo ${{ matrix.folder }}
+      - name: build docker image
+        run: docker build -t $(echo "$(basename "${{ matrix.folder }}"):gpu" | tr '[:upper:]' '[:lower:]') -f ${{ matrix.folder }}/Dockerfile.gpu ${{ matrix.folder }}
```
Comment (tadejsv) [inline .github/workflows/ci.yml:96]: How about this: we set the default working directory using
```yaml
gpu-tests:
  defaults:
    run:
      working-directory: ${{ matrix.folder }}
```
this way we can use `${PWD##*/}` for container name. And we can also use `-f Dockerfile.gpu` as well. I'm not 100% if it would work, but would look much cleaner if it would

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -63,8 +63,58 @@ jobs:
...
+            -d $(echo "$(basename "${{ matrix.folder }}"):gpu" | tr '[:upper:]' '[:lower:]') tail -F /dev/null
+          export DOCKER_EXEC="docker exec $(basename "${{ matrix.folder }}")"
+      - name: run tests in container
+        run: |
+          $DOCKER_EXEC ls
+          $DOCKER_EXEC pip install -r tests/requirements.txt
+          $DOCKER_EXEC pytest -s -v -m gpu tests/
+      - name: stop container
+        if: ${{ always() }}
+        run: docker stop $(docker ps -a -q --filter="name=$(basename "${{ matrix.folder }}")")
```
Comment (tadejsv) [inline .github/workflows/ci.yml]: I think we don't need the filter - as this is the only job running on the runner, I think it is safe to kill all containers.

**Comment by jakob1111996** [comment]
> Make sure to also test the docker image as I am doing in `docker-tests`

Do you mean just starting it with the default entry point?

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -87,29 +87,32 @@ jobs:
...
+        run: docker build -t $(echo "${PWD##*/}:gpu" | tr '[:upper:]' '[:lower:]') -f Dockerfile.gpu .
       - name: start container
         run: |
           docker run \
             --entrypoint "" \
-            --name $(basename ${{ matrix.folder }}) \
+            --name ${PWD##*/} \
             --gpus all \
-            -d $(echo "$(basename "${{ matrix.folder }}"):gpu" | tr '[:upper:]' '[:lower:]') tail -F /dev/null
+            -d $(echo "${PWD##*/}:gpu" | tr '[:upper:]' '[:lower:]') tail -F /dev/null
```
Comment (tadejsv) [inline .github/workflows/ci.yml:106]: Just wondering - why is the `| tr ...` thing needed here?

Comment (tadejsv) [reply .github/workflows/ci.yml:106]: Figured it out - image name needs to be lowercase

**Comment by tadejsv** [comment]
Let's stop at the executors we have now - adding more can be done in other PRs. The main goal of this one is to get the infrastructure for GPU CI/CD in plance, the rest can follow

**Comment by nan-wang** [comment]
> Let's stop at the executors we have now - adding more can be done in other PRs. The main goal of this one is to get the infrastructure for GPU CI/CD in plance, the rest can follow

I am totally agreed on this.

**Comment by nan-wang** [comment]
As for the failing tests on `paddlepaddle`, I think the main issue is the missing dependency of `opencv-python`

**Comment by nan-wang** [comment]
Another issue is about black, https://github.com/jina-ai/executors/pull/128/checks?check_run_id=3464128680#step:7:31

**Comment by jakob1111996** [comment]
Ready to merge - Reviews would be nice

File: jinahub/encoders/audio/VGGISHAudioEncoder/Dockerfile.gpu [hunk 9]
```diff
@@ -0,0 +1,12 @@
+FROM nvidia/cuda:11.3.1-cudnn8-runtime
```
Comment (tadejsv) [inline jinahub/encoders/audio/VGGISHAudioEncoder/Dockerfile.gpu:1]: Make this 11.2, this is max that tensorflow is compatible with
https://www.tensorflow.org/install/source#gpu

File: jinahub/encoders/audio/VGGISHAudioEncoder/Dockerfile.gpu [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+RUN apt update && DEBIAN_FRONTEND="noninteractive" apt install -y python3.8 python3-pip git libsndfile-dev
+RUN JINA_PIP_INSTALL_PERF=1 pip install --no-cache-dir jina==2.0.20
```
Comment (tadejsv) [inline jinahub/encoders/audio/VGGISHAudioEncoder/Dockerfile.gpu:4]:
```suggestion
RUN JINA_PIP_INSTALL_PERF=1 pip install --no-cache-dir jina~=2.0
```

File: jinahub/encoders/audio/VGGISHAudioEncoder/Dockerfile.gpu [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+COPY gpu_requirements.txt gpu_requirements.txt
+RUN pip install --no-cache-dir -r gpu_requirements.txt
+
+COPY . ./vgg_audio_encoder/
+WORKDIR ./vgg_audio_encoder
```
Comment (tadejsv) [inline jinahub/encoders/audio/VGGISHAudioEncoder/Dockerfile.gpu:10]: Let's standardize this everywhere to `/workdir` (no dot)

File: jinahub/encoders/audio/VGGISHAudioEncoder/Dockerfile.gpu (hunk 10 above)
Comment (tadejsv) [reply jinahub/encoders/audio/VGGISHAudioEncoder/Dockerfile.gpu:4]: Also, here you can put this in `gpu_requirements.txt`

File: jinahub/encoders/audio/VGGISHAudioEncoder/tests/integration/test_exec.py [hunk 12]
```diff
@@ -23,11 +24,32 @@ def test_flow_from_yml():
...
+
+@pytest.mark.gpu
+@pytest.mark.docker
+def test_docker_runtime_gpu():
+    with pytest.raises(subprocess.TimeoutExpired):
+        subprocess.run(
+            [
+                'jina',
+                'pea',
+                '--uses=docker://vggishaudioencoder:gpu',
```
Comment (tadejsv) [inline jinahub/encoders/audio/VGGISHAudioEncoder/tests/integration/test_exec.py:47]: Adjust this to the new template (see `CONTRIBUTING.md`)

File: jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt [hunk 13]
```diff
@@ -1,2 +1,3 @@
-git+https://github.com/jina-ai/jina.git@v2.0.0#egg=jina[hub]
-pytest==6.2.4
\ No newline at end of file
+jina~=2.0
```
Comment (tadejsv) [inline jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt:1]:
```suggestion
```

File: jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile [hunk 14]
```diff
@@ -1,7 +1,7 @@
-FROM jinaai/jina:master-py37-perf
+FROM jinaai/jina:2.0.20
```
Comment (tadejsv) [inline jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile:1]:
```suggestion
FROM jinaai/jina:2-py37-perf
```

File: jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile [hunk 15]
```diff
@@ -1,7 +1,7 @@
...
 
 # install git
-RUN apt-get -y update && apt-get install -y git libgomp1 libglib2.0-0 libsm6 libxext6 libfontconfig1 libxrender1
+RUN apt-get -y update && DEBIAN_FRONTEND=noninteractive apt-get install -y git libgomp1 libglib2.0-0 libsm6 libxext6 libfontconfig1 libxrender1 python3-opencv
```
Comment (tadejsv) [inline jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile:4]: Try putting `opencv-python-headless` in requirements instead

**Review by tadejsv — CHANGES_REQUESTED** [review]
Some comments for now, will continue later

File: jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt (hunk 13 above)
Comment (jakob1111996) [reply jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt:1]: We need this. It is in every tests/requirements.txt

Comment (tadejsv) [reply jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt:1]: We agreed to remove it, and in testing CI script you can see that jina gets installed before the requirements

Comment (jakob1111996) [reply jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt:1]: So should I remove it from all tests/requirements.txt ?

Comment (tadejsv) [reply jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt:1]: Yes

File: jinahub/encoders/audio/VGGISHAudioEncoder/tests/conftest.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    subprocess.run(['docker', 'build', '-t', docker_image_name, '.'], check=True)
+    return docker_image_name
+
+
+@pytest.fixture(scope='session')
+def build_docker_image_gpu(docker_image_name: str) -> str:
+    image_name = f'{docker_image_name}:gpu'
+    subprocess.run(
+        ['docker', 'build', '-t', image_name, '-f', 'Dockerfile.gpu', '.'], check=True
+    )
```
Comment (tadejsv) [inline jinahub/encoders/audio/VGGISHAudioEncoder/tests/conftest.py:23]:
```suggestion
    )
    return image_name
```

File: jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile.gpu [hunk 17]
```diff
@@ -0,0 +1,20 @@
+FROM nvidia/cuda:10.1-cudnn7-runtime
+
+RUN apt update && DEBIAN_FRONTEND=noninteractive apt install -y python3.8 python3.8-distutils git libgomp1 libglib2.0-0 libsm6 libxext6 libfontconfig1 libxrender1 wget
+RUN wget https://bootstrap.pypa.io/get-pip.py
+RUN python3.8 get-pip.py
```
Comment (tadejsv) [inline jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile.gpu:5]: Why not just add `python3-pip` ?

File: jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile.gpu [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+RUN pip3 install --no-cache-dir -r gpu_requirements.txt
+
+COPY . /workdir/
+WORKDIR /workdir
+
+ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/lib/x86_64-linux-gnu/:$LD_LIBRARY_PATH
+RUN ln -s /usr/lib/x86_64-linux-gnu/libcudnn.so.7 /usr/lib/x86_64-linux-gnu/libcudnn.so
+SHELL ["/bin/bash", "-c"]
+RUN for i in $(find /usr/local/cuda/lib64/ -iname "*.so.10"); do ln -s $i ${i::-3} ; done
+RUN ln -s /usr/lib/x86_64-linux-gnu/libcublas.so.10 /usr/lib/x86_64-linux-gnu/libcublas.so
```
Comment (tadejsv) [inline jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile.gpu:18]: Seems complicated. Since paddle is just a copy of pytorch, can you check if it works like it too, so no CUDA base image required? On the installation page it seems it would

File: jinahub/encoders/image/ImagePaddlehubEncoder/tests/unit/test_encoder.py [hunk 19]
```diff
@@ -41,3 +43,30 @@ def dist(a, b):
     assert small_distance < dist('banana2', 'studio')
     assert small_distance < dist('airplane', 'studio')
     assert small_distance < dist('airplane', 'satellite')
+
+
+@pytest.mark.gpu
+def test_imagepaddlehubencoder_encode_gpu(test_images: Dict[str, np.array]):
```
Comment (tadejsv) [inline jinahub/encoders/image/ImagePaddlehubEncoder/tests/unit/test_encoder.py:49]: Too complidated, just do a simple encoding test

File: jinahub/encoders/image/ImageTFEncoder/Dockerfile [hunk 20]
```diff
@@ -1,4 +1,4 @@
-FROM jinaai/jina:2.0
+FROM jinaai/jina:2.0.20
```
Comment (tadejsv) [inline jinahub/encoders/image/ImageTFEncoder/Dockerfile:1]:
```suggestion
FROM jinaai/jina:2-py37-perf
```

File: jinahub/encoders/image/ImageTorchEncoder/Dockerfile.gpu [hunk 21]
```diff
@@ -0,0 +1,12 @@
+FROM nvidia/cuda:11.3.1-cudnn8-runtime
```
Comment (tadejsv) [inline jinahub/encoders/image/ImageTorchEncoder/Dockerfile.gpu:1]: Not needed, torch can work with jina base image on GPU

File: jinahub/encoders/image/ImageTorchEncoder/tests/unit/test_image_torch_encoder.py [hunk 22]
```diff
@@ -54,6 +53,26 @@ def test_encode_image_returns_correct_length(
...
 
+@pytest.mark.parametrize(
+    'traversal_paths, docs',
+    [
+        (('r',), pytest.lazy_fixture('docs_with_blobs')),
+        (('c',), pytest.lazy_fixture('docs_with_chunk_blobs')),
+    ],
+)
+@pytest.mark.gpu
+def test_encode_image_returns_correct_length_gpu(
```
Comment (tadejsv) [inline jinahub/encoders/image/ImageTorchEncoder/tests/unit/test_image_torch_encoder.py:64]: Too complicated, just do a simple encoding

**Review by tadejsv — COMMENTED** [review]
More comments (all for this round)

File: jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile.gpu (hunk 17 above)
Comment (jakob1111996) [reply jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile.gpu:5]: This did not work in the container. It always uninstalled python3.8 and setup 3.6 instead. It has somethind to do with the cuda container.... This way it works

File: jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile.gpu (hunk 18 above)
Comment (jakob1111996) [reply jinahub/encoders/image/ImagePaddlehubEncoder/Dockerfile.gpu:18]: I really tried around a lot with paddle.... Nothing seemed to work except this weird workaround here.

File: jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt (hunk 13 above)
Comment (tadejsv) [reply jinahub/encoders/audio/VGGISHAudioEncoder/tests/requirements.txt:1]: This still needs to be removed

File: jinahub/encoders/image/ImageTFEncoder/tests/requirements.txt [hunk 23]
```diff
@@ -1,4 +1,4 @@
-git+https://github.com/jina-ai/jina.git@v2.0.3#egg=jina[hub]
+jina~=2.0
```
Comment (tadejsv) [inline jinahub/encoders/image/ImageTFEncoder/tests/requirements.txt:1]: Remove
```suggestion
```

File: jinahub/encoders/image/ImageTorchEncoder/Dockerfile [hunk 24]
```diff
@@ -1,4 +1,4 @@
-FROM jinaai/jina:2.0.3
+FROM jinaai/jina:2.0.20
```
Comment (tadejsv) [inline jinahub/encoders/image/ImageTorchEncoder/Dockerfile:1]:
```suggestion
FROM jinaai/jina:2-py37-perf
```

File: jinahub/encoders/image/ImageTorchEncoder/torch_encoder.py [hunk 25]
```diff
@@ -36,20 +35,27 @@ class ImageTorchEncoder(Executor):
...
     ):
         super().__init__(*args, **kwargs)
+        self.logger = JinaLogger(self.__class__.__name__)
+        if (device not in ['cpu', 'cuda']) and (not device.startswith('cuda:')):
+            self.logger.error(
+                f'Torch device {device} not supported. Must be cpu or cuda!'
+            )
+            raise RuntimeError(
+                f'Torch device {device} not supported. Must be cpu or cuda!'
+            )
```
Comment (tadejsv) [inline jinahub/encoders/image/ImageTorchEncoder/torch_encoder.py:57]: One last thing: this can be removed :)

**Review by tadejsv — APPROVED** [review]
Smash that merge button!

**Comment by jakob1111996** [comment]
Party hard
